packages/babel-core/src/config/plugin.js

```diff
@@ -1,43 +1,123 @@
+    $PropertyType<PluginObject, "parserOverride">,
+  >),
+  generatorOverride: (assertFunction: Validator<
+    $PropertyType<PluginObject, "generatorOverride">,
+  >),
+};
+
+function assertVisitorMap(key: string, value: mixed): VisitorMap {
+  const obj = assertObject(key, value);
+  if (obj) {
+    Object.keys(obj).forEach(prop => assertVisitorHandler(prop, obj[prop]));
+
+    if (obj.enter || obj.exit) {
       throw new Error(
-        "Plugin .manipulateOptions must be a function, null, or undefined",
+        `.${key} cannot contain catch-all "enter" or "exit" handlers. Please target individual nodes.`,
       );
     }
-    if (plugin.post != null && typeof plugin.post !== "function") {
-      throw new Error("Plugin .post must be a function, null, or undefined");
-    }
-    if (plugin.pre != null && typeof plugin.pre !== "function") {
-      throw new Error("Plugin .pre must be a function, null, or undefined");
-    }
-    if (plugin.visitor != null && typeof plugin.visitor !== "object") {
-      throw new Error("Plugin .visitor must be an object, null, or undefined");
-    }
+  }
+  return (obj: any);
+}
+
+function assertVisitorHandler(
+  key: string,
+  value: mixed,
+): VisitorHandler | void {
+  if (value && typeof value === "object") {
+    Object.keys(value).forEach(handler => {
+      if (handler !== "enter" && handler !== "exit") {
+        throw new Error(
+          `.visitor["${key}"] may only have .enter and/or .exit handlers.`,
+        );
+      }
+    });
+  } else if (typeof value !== "function") {
+    throw new Error(`.visitor["${key}"] must be a function`);
+  }
+
+  return (value: any);
+}
+
+type VisitorHandler = Function | { enter?: Function, exit?: Function };
+export type VisitorMap = {
+  [string]: VisitorHandler,
+};
 
+export type PluginObject = {
+  name?: string,
+  manipulateOptions?: Function,
+
+  pre?: Function,
+  post?: Function,
+
+  inherits?: Function,
+  visitor?: VisitorMap,
+
+  parserOverride?: Function,
+  generatorOverride?: Function,
+};
+
+export function validatePluginObject(obj: {}): PluginObject {
+  Object.keys(obj).forEach(key => {
+    const validator = VALIDATORS[key];
+
+    if (validator) validator(key, obj[key]);
+    else throw new Error(`.${key} is not a valid Plugin property`);
+  });
+
+  return (obj: any);
+}
+
+export default class Plugin {
+  key: ?string;
+  manipulateOptions: Function | void;
+  post: Function | void;
+  pre: Function | void;
+  visitor: {};
+
+  parserOverride: Function | void;
+  generatorOverride: Function | void;
+
+  options: {};
+
+  constructor(plugin: PluginObject, options: {}, key?: string) {
     this.key = plugin.name || key;
 
     this.manipulateOptions = plugin.manipulateOptions;
     this.post = plugin.post;
     this.pre = plugin.pre;
-    this.visitor = plugin.visitor;
+    this.visitor = plugin.visitor || {};
+    this.parserOverride = plugin.parserOverride;
+    this.generatorOverride = plugin.generatorOverride;
```

Comment:
We really need to start documenting this stuff now lol

Reply:
I don't really mind that it's not super well publicized, I just want a clear API for it :P

Reply:
test

Reply:
sdfds
```